src/Lib.P2P/Cryptography/CtrStreamCipher.cs

```diff
@@ -0,0 +1,145 @@
+        private readonly byte[] counter;
+        private readonly byte[] counterOut;
+        private int byteCount;
+        private byte[] IV;
+
+        /// <summary>
+        ///   Creates a new instance of the <see cref="CtrStreamCipher"/> with
+        ///   the specified cipher.
+        /// </summary>
+        /// <param name="cipher">
+        ///   The cipher to produce the output counter.  Typically
+        ///   <see cref="Org.BouncyCastle.Crypto.Engines.AesEngine"/>.
+        /// </param>
+        public CtrStreamCipher(IBlockCipher cipher)
+        {
+            this.cipher = cipher;
+            blockSize = cipher.GetBlockSize();
+            counter = new byte[blockSize];
+            counterOut = new byte[blockSize];
+            IV = new byte[blockSize];
+        }
+
+        /// <summary>
+        ///   The name of this algorithm.
+        /// </summary>
+        public string AlgorithmName => cipher.AlgorithmName + "/CTR";
+
+        /// <summary>
+        ///   Init the cipher.
+        /// </summary>
+        /// <param name="forEncryption">
+        ///   Ignored.
+        /// </param>
+        /// <param name="parameters">
+        ///   Must be a <see cref="ParametersWithIV"/>.
+        /// </param>
+        /// <example>
+        /// var encrypt = new CtrStreamCipher(new AesEngine());
+        /// var p = new ParametersWithIV(new KeyParameter(key), iv);
+        /// encrypt.Init(true, p);
+        /// </example>
+        public void Init(bool forEncryption, ICipherParameters parameters)
+        {
+            var ivParam = parameters as ParametersWithIV;
+            if (ivParam == null)
+                throw new ArgumentException("CTR mode requires ParametersWithIV", "parameters");
+
+            IV = Arrays.Clone(ivParam.GetIV());
+
+            if (blockSize < IV.Length)
+                throw new ArgumentException("CTR mode requires IV no greater than: " + blockSize + " bytes.");
+
+            var maxCounterSize = Math.Min(8, blockSize / 2);
+            if (blockSize - IV.Length > maxCounterSize)
+                throw new ArgumentException("CTR mode requires IV of at least: " + (blockSize - maxCounterSize) +
+                    " bytes.");
+
+            // if null it's an IV changed only.
+            if (ivParam.Parameters != null) cipher.Init(true, ivParam.Parameters);
+
+            Reset();
+        }
+
+        /// <inheritdoc />
+        public void Reset()
+        {
+            byteCount = 0;
+            Arrays.Fill(counter, (byte) 0);
+            Array.Copy(IV, 0, counter, 0, IV.Length);
+            cipher.Reset();
+        }
+
+        /// <inheritdoc />
+        public void ProcessBytes(byte[] input, int inOff, int length, byte[] output, int outOff)
+        {
+            if (outOff + length > output.Length) throw new DataLengthException("Output buffer too short");
+
+            if (inOff + length > input.Length) throw new DataLengthException("Input buffer too small");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)